web/src/LogStore.ts

```diff
@@ -243,57 +264,64 @@ class LogStore {
         continue
       }
 
-      let currentLine = {
-        manifestName: span.manifestName,
-        text: segment.text,
-        level: segment.level,
+      let line = this.lineCache[segment.globalLineIndex]
+      if (!line) {
+        line = this.createLogLineHelper(i, segment, span)
+        this.lineCache[segment.globalLineIndex] = line
       }
-      isFirstLine = false
+      result.push(line)
+    }
 
-      // If this segment is not complete, run ahead and try to complete it.
-      if (segment.isComplete()) {
-        lastLineCompleted = true
+    return result
+  }
+
+  private createLogLineHelper(
+    i: number,
+    segment: LogSegment,
+    span: LogSpan
+  ): LogLine {
+    let spanId = segment.spanId
+    let currentLine = {
+      manifestName: span.manifestName,
+      text: segment.text,
+      level: segment.level,
+    }
+
+    // If this segment is not complete, run ahead and try to complete it.
```

Comment:
i'm confused why a comment about "if segment is not complete" is right above "if segment.IsComplete()..." -- this comment seems to refer to different code, and should be nearer to the code it refers to

Reply:
good question! there's an early exit on line 297. I'll clarify the comment a bit